### PR TITLE
prov/shm: refactor address mapping and smr_cmd structure + coverity fix

### PR DESCRIPTION
**Pull request by aingerson**
- Change format of address storage so allow peers to exchange addresses through the shared memory region
- Create specific smr_cmd over ofi_cmd
- Increase msg_id to uint64_t and hold context
- Some renaming
- Fix inconsistent locking order reported by coverity

Signed-off-by: aingerson <alexia.ingerson@intel.com>

### CODE REVIEW

File: prov/shm/src/smr_ep.c [hunk 1]
```diff
@@ -167,13 +167,16 @@ void smr_progress_resp(struct smr_ep *ep)
...
-	while (!ofi_cirque_isempty(smr_resp_queue(ep->region))) {
+	fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
+	while (!ofi_cirque_isempty(smr_resp_queue(ep->region)) &&
+	       !ofi_cirque_isfull(ep->util_ep.tx_cq->cirq)) {
 		resp = ofi_cirque_head(smr_resp_queue(ep->region));
-		if (resp->status)
+		if (resp->status == FI_EBUSY)
 			break;
-		ep->tx_comp(ep, NULL, resp->flags);
+		ep->tx_comp(ep, (void *) resp->msg_id, FI_SEND | resp->flags);
```
Comment (shefty) [inline prov/shm/src/smr_ep.c]: Are you sure that this is a send completion, versus rma or atomic?

File: prov/shm/src/smr_ep.c [hunk 2]
```diff
@@ -185,9 +188,11 @@ static int smr_progress_inline(struct smr_cmd *cmd, struct smr_ep_entry *entry)
 	if (ret != cmd->hdr.op.size) {
 		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 			"recv truncated");
-		return -FI_EIO;
+		entry->resv = FI_EIO;
```
Comment (shefty) [inline prov/shm/src/smr_ep.c]: If you're using the resv field to store a value then it's not really reserved.

File: prov/shm/src/smr_ep.c (hunk 1 above)
Comment (aingerson) [reply prov/shm/src/smr_ep.c]: For now, yes. When rma and atomics are implemented, no.

File: prov/shm/src/smr_ep.c (hunk 2 above)
Comment (aingerson) [reply prov/shm/src/smr_ep.c]: Ahh, in my head that was standing for "result value". I'll change the name.

File: prov/shm/src/smr_ep.c [hunk 3]
```diff
@@ -290,23 +296,23 @@ void smr_progress_cmd(struct smr_ep *ep)
...
 			break;
 		default:
 			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 				"unidentified operation type\n");
-			ret = -FI_EINVAL;
+			entry->resv = FI_EINVAL;
 		}
-		ep->rx_comp(ep, entry->context, 0, ret, NULL, &addr,
-			    cmd->hdr.op.tag);
+		ep->rx_comp(ep, entry->context, FI_RECV | entry->flags,
```
Comment (shefty) [inline prov/shm/src/smr_ep.c]: You may be able to set FI_RECV at the lowest layer -- i.e. wherever rx_comp() points at.

File: prov/util/src/util_shm.c [hunk 4]
```diff
@@ -43,22 +43,29 @@
 #include <fi_shm.h>
 
 
+static void smr_peer_addr_init(struct smr_addr *peer)
+{
+	memset(peer->name, '\0', SMR_NAME_SIZE);
```
Comment (shefty) [inline prov/util/src/util_shm.c]: nit: memset wants an int as the 2nd arg, not a char.  So just 0 works better here.

**Pull request merged**